36-vending-machine-monitor/firmware/vending-machine/vending-machine.ino

```diff
@@ -0,0 +1,1125 @@
+ * @param oldValue  The previous value of the variable.
+ * @param newValue  The current value of the variable.
+ * @return J*       The notification that was added.
+ */
+J* addEnvironmentChangedNotification(J* changed, const char* name, J* oldValue, J* newValue) {
+    J* item = JCreateObject();
+    if (oldValue) {
+        JAddItemToObject(item, "old_value", oldValue);
+    }
+    if (newValue) {
+        JAddItemToObject(item, "new_value", newValue);
+    }
+    JAddItemToObject(changed, name, item);
+    return item;
+}
+
+J* addEnvironmentChangedNotification(J* changed, const char* name, double oldValue, double newValue) {
+    return addEnvironmentChangedNotification(changed, name, JCreateNumber(oldValue), JCreateNumber(newValue));
+}
+
+J* notifyEnvironmentError(J* errors, const char* name, const char* text, const char* value) {
+    J* item = JCreateObject();
+    JAddStringToObject(item, "error", text);
+    JAddStringToObject(item, "value", value);
+    JAddItemToObject(errors, name, item);
+    return item;
+}
+
+/**
+ * @brief Updates an integer variable from the value in the environment.
+ * 
+ * @param env       The environment variables
+ * @param changed   Receives notifications of variables that changed.
+ * @param error     Receives notificiations of environment errors.
+ * @param result    The variable to store the value
+ * @param envVar    The name of the environment variable to read.
+ * @param multiplier A multiplier to use after fetching the value from the environment. This is useful
+ *  when the variable value and internal value use different units. (Such as seconds vs milliseconds.)
+ * @param defaultValue The default value to use when the value is not defined in the environment.
+ */
+void setIntValueFromEnvironment(J* env, J* changed, J* error, uint32_t* result, const char* envVar, uint32_t multiplier, uint32_t defaultValue) {
+    const char* value_str = JGetString(env, envVar);
+    uint32_t value = defaultValue;
+    if (value_str && *value_str) {    // variable is defined
+        char *endptr;
+        long env_value = strtol(value_str, &endptr, 10);
+        if (*endptr || env_value<0) {
+            notifyEnvironmentError(error, envVar, "not a valid whole positive number.", value_str);
+        }
+        else {
+            value = env_value * multiplier;
+            if (*result!=value) {
+                addEnvironmentChangedNotification(changed, envVar, *result/multiplier, value);
+            }
+        }
+    }
+    *result = value;
+}
+
+// placeholder variable names for later interpolation
+#define ENV_COL_NAME_FMT "col_%d_name"
+#define ENV_COL_EMPTY_FMT "col%sempty_mm"
+#define ENV_COL_LOW_FMT "col%slow_mm"
+#define ENV_COL_FULL_FMT "col%sfull_mm"
+#define ENV_COL_ITEM_FMT "col%sitem_mm"
+
+/**
+ * @brief The default column sizes defined by the environment.
+ */
+SodaStack defaultColumnSize;
+const SodaStack INITIAL_STACK_SIZE; // zero as the default
```

Comment:
I would add = 0 here, even if it's initialized to zero automatically.

Reply:
I can't add = 0 or = { 0 } because it has a default constructor that initializes the values to zero.

```
/Users/mat/b/nifty-fifty/36-vending-machine-monitor/firmware/vending-machine/vending-machine.ino:877:40: error: could not convert '{0}' from '<brace-enclosed initializer list>' to 'const SodaStack'
  877 | const SodaStack INITIAL_STACK_SIZE = {0};
      |                                        ^
      |                                        |
      |                                        <brace-enclosed initializer list>
     ```

Reply:
Ah, never mind then.